optimus.py

```diff
@@ -0,0 +1,798 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+__license__ = "GPL-2.0"
+__author__ = "Martin Kurbanov"
+__email__ = "mmkurbanov@salutedevices.com"
+__copyright__ = "Copyright (c) 2023, SaluteDevices"
+__version__ = '0.0.1'
+
+try:
+    from pyamlboot.pyamlboot import pyamlboot
+except ImportError:
+    from pyamlboot import pyamlboot
+
+import logging
+import time
+import typing
+from collections import OrderedDict
+from dataclasses import dataclass
+from struct import pack, unpack
+
+import usb_backend
+
+USB_BACKEND = usb_backend.get_backend()
+
+
+class BulkCmdError(Exception):
+    pass
+
+
+class TplCmdError(Exception):
+    pass
+
+
+class SocId:
+    STAGE_MINOR_IPL = 0   # Initial Program Loader
+    STAGE_MINOR_SPL = 8   # Secondary Program Loader
+    STAGE_MINOR_TPL = 16  # Tertiary Program Loader
+
+    def __init__(self, socid):
+        self._raw = socid
+        self.major = ord(socid[0])
+        self.minor = ord(socid[1])
+        self.stage_major = ord(socid[2])
+        self.stage_minor = ord(socid[3])
+        if len(socid) > 4:
+            self.need_password = bool(ord(socid[4]))
+            self.password_ok = bool(ord(socid[5]))
```

Comment:
Instead of directly assigning on creation, why not create getter properties for these values?

Something like this:

```
class SocId:

    @property
    def major(self):
        return ord(self._raw[0])
```